File: chatly-backend/src/main/java/com/chatly/model/mongo/Story.java
package com.chatly.model.mongo;

import com.chatly.model.enums.StoryPrivacy;
import com.chatly.model.enums.StoryType;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Document(collection = "stories")
@CompoundIndexes({
        @CompoundIndex(def = "{'userId': 1, 'createdAt': -1}")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Story {

    @Id
    private String id;

    private String userId;

    private StoryType type;

    private String content;

    private String mediaUrl;

    private String musicUrl;

    private String musicName;

    private Integer bgIndex;

    private Integer fontSize;

    @Builder.Default
    private StoryPrivacy privacy = StoryPrivacy.EVERYONE;

    /** IDs of users who have viewed this story. */
    @Builder.Default
    private List<String> viewerIds = new ArrayList<>();

    @Builder.Default
    private int viewCount = 0;

    @CreatedDate
    @Indexed(expireAfterSeconds = 86400)
    private Instant createdAt;
}

File: chatly-frontend/src/types/story.ts
export type StoryType = "TEXT" | "PHOTO" | "VIDEO";

export const StoryType: Record<StoryType, StoryType> = {
    TEXT: "TEXT",
    PHOTO: "PHOTO",
    VIDEO: "VIDEO",
};

export type StoryPrivacy =
    | "EVERYONE"
    | "FRIENDS_ONLY"
    | "CLOSE_FRIENDS"
    | "ONLY_ME";

export const StoryPrivacy: Record<StoryPrivacy, StoryPrivacy> = {
    EVERYONE: "EVERYONE",
    FRIENDS_ONLY: "FRIENDS_ONLY",
    CLOSE_FRIENDS: "CLOSE_FRIENDS",
    ONLY_ME: "ONLY_ME",
};

export interface StoryUser {
    id: string;
    displayName?: string;
    avatarUrl?: string;
    username?: string;
}

export interface Story {
    id: string;
    userId: string;
    type: StoryType;
    content?: string;
    mediaUrl?: string;
    musicUrl?: string;
    musicName?: string;
    bgIndex?: number;
    fontSize?: number;
    privacy: StoryPrivacy;
    viewCount?: number;
    viewedByMe?: boolean;
    createdAt: string;
    user?: StoryUser;
}

export interface StoryCreationRequest {
    type: StoryType;
    content?: string;
    mediaUrl?: string;
    musicUrl?: string;
    musicName?: string;
    bgIndex?: number;
    fontSize?: number;
    privacy: StoryPrivacy;
}

export interface StoryReactionResponse {
    id: string;
    storyId: string;
    userId: string;
    emoji: string;
    createdAt: string;
    user?: StoryUser;
}

export interface StoryReplyResponse {
    id: string;
    storyId: string;
    userId: string;
    content: string;
    createdAt: string;
    user?: StoryUser;
}
